File: typescript/maths.ts
module Maths {
  export interface Expression {
    evaluate() : number
  }

  export class Constant implements Expression {
    n: number;

    constructor(n: number) {
      this.n = n;
    }

    evaluate() : number {
      return this.n;
    }
  }

  abstract class BinaryOp implements Expression {
    left: Expression;
    right: Expression;

    constructor(left: Expression, right: Expression) {
      this.left = left;
      this.right = right;
    }

    abstract op(x: number, y: number) : number;

    evaluate() : number {
      return this.op(this.left.evaluate(), this.right.evaluate());
    }
  }

  export class Sum extends BinaryOp {
    op(x: number, y: number) : number {
      return x + y;
    }
  }

  export class Difference extends BinaryOp {
    op(x: number, y: number) : number {
      return x - y;
    }
  }

  export class Product extends BinaryOp {
    op(x: number, y: number) : number {
      return x * y;
    }
  }

  export class Quotient extends BinaryOp {
    op(x: number, y: number) : number {
      return Math.floor(x / y);
    }
  }

  export class Modulo extends BinaryOp {
    op(x: number, y: number) : number {
      return x % y;
    }
  }

  export class Power extends BinaryOp {
    op(x: number, y: number) : number {
      return Math.pow(x, y);
    }
  }
}

File: csharp/Maths.cs
namespace Math
{
    public interface Expression {
        int Value { get; }
    }

    public sealed class Constant : Expression {
        public int Value { get; set; }
    }

    public abstract class BinaryOp : Expression {
        Expression x;
        Expression y;

        public BinaryOp(Expression x, Expression y) {
            this.x = x;
            this.y = y;
        }

        public int Value => Op(x.Value, y.Value);

        protected abstract int Op(int x, int y);
    }

    public sealed class Sum : BinaryOp {
        public Sum(Expression x, Expression y)
        : base(x, y) {
        }

        protected override int Op(int x, int y) => x + y;
    }

    public sealed class Difference : BinaryOp {
        public Difference(Expression x, Expression y)
        : base(x, y) {
        }

        protected override int Op(int x, int y) => x - y;
    }

    public sealed class Product : BinaryOp {
        public Product(Expression x, Expression y)
        : base(x, y) {
        }

        protected override int Op(int x, int y) => x * y;
    }

    public sealed class Quotient : BinaryOp {
        public Quotient(Expression x, Expression y)
        : base(x, y) {
        }

        protected override int Op(int x, int y) => x / y;
    }

    public sealed class Power : BinaryOp {
        public Power(Expression x, Expression y)
        : base(x, y) {
        }

        protected override int Op(int x, int y) => (int)System.Math.Pow(x, y);
    }

    public sealed class Modulo : BinaryOp {
        public Modulo(Expression x, Expression y)
        : base(x, y) {
        }

        protected override int Op(int x, int y) => x % y;
    }
}